File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        leftPrefix = [1] * len(nums)
        rightSuffix = [1] * len(nums)
        currProd = 1

        for i in range(0,len(nums)-1):
            currProd *= nums[i]
            leftPrefix[i+1] *= currProd
        
        currProd = 1
        for i in range(len(nums)-1,0,-1):
            currProd *= nums[i]
            rightSuffix[i-1] *= currProd

        res = [1] * len(nums)
        for i in range(len(leftPrefix)):
            res[i] = rightSuffix[i] * leftPrefix[i]
        return res

File: 238-product-of-array-except-self/product-of-array-except-self.go
func productExceptSelf(nums []int) []int {
    prefixProd := 1

    // Init 
    n := len(nums)
    res := make([]int, n)
    for i := range n {
        res[i] = 1
    }

    // Calc left prefix arr
    for i,val := range nums[0:n-1] {
        prefixProd *= val
        res[i+1] = prefixProd
    }

    // Calc right suffix arr
    prefixProd = 1
    for i := n-1; i>=1; i-- {
        prefixProd *= nums[i]
        res[i-1] *= prefixProd
    }
    return res

}